classpath/java/io/BufferedInputStream.java

```diff
@@ -54,28 +59,7 @@ public int read(byte[] b, int offset, int length) throws IOException {
 
       count += remaining;
       position += remaining;
-      offset += remaining;
-      length -= remaining;
     }
-
-    while (length > 0) {
```

Comment:
Careful, because that's a deviation from the spec: "This method blocks until input data is available" (http://docs.oracle.com/javase/7/docs/api/java/io/InputStream.html#read(byte[])).

Agreed - that may not be the most efficient (for certain scenarios, anyway) - but it is what we need to be doing here.

If you're set on using non-blocking communication (which, by the way, is not a panacea, for either performance or code complexity), you should switch to using the java.nio.channels.SocketChannel, which is non-blocking.